Add chapters to video transcripts

nicolevanderhoeven opened the pull request:
This PR adds chapters generation to the video transcript Python script originally added [here](https://github.com/open-telemetry/sig-end-user/pull/134), adding a new function to use OpenAI to create chapters (timestamped sections of the video). When added to the YouTube description of a video, these chapters allow viewers to skip to relevant portions of the video and also give everyone a better idea of what the video entails.

In addition, I made the following improvements to video transcription:
- I upgraded the youtube-transcript-api library to the latest one, since I noticed that the old one was no longer working with the current YouTube API
- I added some delays to prevent rate limiting (which I ran into while testing the transcript + chapter generation). I also added some error handling to detect things like IP block detection and XML parse errors. The rate limiting improvements include retries with an exponential backoff.
- I added these changes to the documentation and clarified how to use this script.
- I added an `env.example` to show what the `.env` file should look like.

nicolevanderhoeven committed:
Add chapters/timestamps feature and improve API reliability

- Add create_chapters() function to generate YouTube-style timestamps using OpenAI
- Integrate chapters section into markdown output between summary and transcript
- Add comprehensive rate limiting to avoid YouTube API quota issues
- Implement get_video_transcript_with_retry() with exponential backoff
- Add robust error handling for quota exceeded and API failures
- Improve transcript validation and filtering ([Music], [Applause], etc)
- Fix Japanese language code from 'jp' to 'ja'
- Increase batch size from 10 to 50 for better efficiency
- Add progress indicators and better logging throughout

nicolevanderhoeven committed:
Update README to document chapters feature and improvements

- Document new chapters/timestamps generation feature
- Add comprehensive usage examples with command-line options
- Document output format and file structure
- Add Features & Reliability section covering rate limiting and error handling
- Clarify OpenAI API key requirements and AI-enhanced features
- Document multi-language transcript support

nicolevanderhoeven committed:
Fix pydantic dependency conflict in requirements.txt

Update pydantic_core from 2.39.0 to 2.33.2 to match the version
required by pydantic 2.11.9, resolving pip installation error.

nicolevanderhoeven committed:
Add env.example

nicolevanderhoeven committed:
Update README with instructions to cp env.example

nicolevanderhoeven committed:
Fix transcript fetching by upgrading youtube-transcript-api

Major changes:
- Upgrade youtube-transcript-api from 0.6.3 to 1.2.3 (fixes empty response issue)
- Update code to use new 1.x API (YouTubeTranscriptApi().fetch())
- Remove deprecated fallback to old static methods
- Enhance error handling for 429 rate limits with 60-120s delays
- Detect XML parse errors as potential rate limiting
- Increase max retries from 3 to 5 attempts

Root cause: YouTube changed their API and version 0.6.3 was returning
empty responses (not rate limiting). The library can now successfully
fetch transcripts and generate chapters.

nicolevanderhoeven committed:
Optimize rate limiting delays for faster processing

Reduced unnecessary delays now that transcript API is fixed:
- Remove initial 5-15s startup delay
- Reduce pagination delays from 3-8s/5-15s to 1-3s
- Reduce inter-video delays from 10-30s to 2-5s
- Reduce API call separation from 10-20s to 2-4s

Keep essential protections:
- YouTube Data API quota error handling (60s retry)
- 429 rate limit detection and handling (60-120s retry)
- XML parse error detection
- Small delays to avoid API hammering

Result: ~3-5x faster processing while maintaining API safety.

nicolevanderhoeven committed:
Add detection and guidance for YouTube IP blocking

- Detect IP block errors specifically (vs rate limiting)
- Stop retries immediately when IP is blocked (no point retrying)
- Add comprehensive troubleshooting section to README
- Provide clear workaround options for users
- Import TooManyRequests exception for better error handling

IP blocks are different from rate limits and require different solutions
like waiting 24-48 hours, switching networks, or using cookie auth.

nicolevanderhoeven committed:
Delete old transcripts and regenerate with timestamps

reese-lee reviewed (approved):
Thank you for doing this. It's interesting to see the differences between some of the old AI-generated transcripts and the new one. 

I think it works in general for most videos, but I did notice that with the Humans of OTel interviews, it summarizes everyone's thoughts instead of leaving them individual, whereas the whole point of a video like that IS to showcase the individuals.

nicolevanderhoeven committed:
Merge branch 'main' into main

nicolevanderhoeven committed:
Strip AI preamble from chapter generation

- Add clean_ai_preamble() function to remove conversational preamble lines
- Remove phrases like 'Sure! Here are the key moments...' from chapter output
- Preserve only actual timestamp lines in generated transcripts

nicolevanderhoeven committed:
Add timestamp insertion and validation to transcript generation

- Insert chapter timestamps inline in cleaned transcripts at matching text positions
- Add video duration constraint to prevent AI from generating timestamps beyond video length
- Improve transcript cleanup to preserve exact wording and speaker names
- Add post-processing validation to verify chapter timestamps match content
- Limit timestamp corrections to ±60 second window around original time
- Use window-based matching with chapter titles for accurate timestamp placement
- Lower AI temperature for more accurate and deterministic results

nicolevanderhoeven committed:
Enforce 10-chapter maximum limit in chapter generation

- Move chapter limit to CRITICAL INSTRUCTIONS section
- Use stronger mandatory language (MUST, NO MORE than 10)
- Add reminder at end of prompt to reinforce limit
- Prevents AI from generating 19-24 chapters per video

nicolevanderhoeven committed:
Regenerated transcripts

nicolevanderhoeven committed:
Simplify timestamp window creation in find_correct_timestamp

- Replace segment-based windowing with simpler 10-second intervals
- Remove complex window_key logic and seen_times tracking
- Create windows directly at 10-second intervals within search range
- Makes timestamp finding more predictable and easier to debug

nicolevanderhoeven committed:
Add unit tests for transcript processing

- Test timeline building with various intervals and edge cases
- Test filtering of [Music] and [Applause] markers
- Test timestamp parsing and formatting utilities
- Test roundtrip conversions and precision handling

nicolevanderhoeven committed:
Add Python virtual environment entries to .gitignore

- Add venv/, .venv/, env/, ENV/ directories
- Add .env file (for API keys/secrets)
- Prevents accidentally committing large virtual environments

nicolevanderhoeven commented:
@reese-lee Thanks for reviewing! I've just done another pass to address your comments.

> * Is it possible to remove the line the AI outputs prior to the summarized transcripts, e.g., "Sure! Here are the key moments identified from the transcript"

Sure is! I've just added it [here](https://github.com/open-telemetry/sig-end-user/pull/200/commits/d697275b35dd021db421a04002305cdb2a1d45d6).

> * Is is possible to have the chapter titles added within the summarized transcript itself at the appropriate sections? Instead of just at the beginning?

Oops, I just put in some logic to add the timestamps of the chapters within the transcript itself, but now I'm rereading this and thinking that you meant the descriptions, not just the timestamps. Before I change this, I just wanted to double-check what you're asking for here. 

Currently ([here's an example of a generated transcript](https://github.com/open-telemetry/sig-end-user/blob/550cfaaf3ffa7d4b3209ffd65e7a0555b3a40b08/video-transcripts/transcripts/2025-11-13T08:33:00Z-humans-of-otel-live-from-kubecon-na-2025.md), there is a chapter at the beginning (`00:27:20 Guest introduction: Diana`) with both the timestamp and description, and then in the transcript, the time is also inserted there (`[00:27:20] **Reese:** Diana, welcome. And thank you guys.`

Would you prefer for that line to look like:
```markdown
### Guest introduction: Diana
**Reese:** Diana, welcome. And thank you guys.
```

instead? Do you want the timestamp there at that point too, or just the heading for the chapter description?

danielgblanco commented on video-transcripts/transcripts.py:
```diff
@@ -22,41 +32,73 @@ def get_playlist_videos(playlist_id):
+            if not next_page_token:
+                break
+                
+            # Small delay between pagination requests
```
Is this delay, and the 200ms delay below needed? Would this not be caught by the 403 exception?

nicolevanderhoeven committed:
Add ## Transcript heading before cleaned transcript section

nicolevanderhoeven committed:
Improve chapter timestamp accuracy by using raw transcript timeline

- Add timeline skeleton to chapter generation prompt showing actual timestamps every 30s
- Remove skip logic for 00:00:00 chapter heading in transcript insertion
- Reuse existing time-to-text mapping logic for consistency
- Fixes issue where chapter timestamps could be off by several minutes

nicolevanderhoeven committed:
Improve chapter selection using summary and full video coverage

- Pass video summary to chapter generation to guide topic selection
- Add explicit guidance to skip small talk not mentioned in summary
- Fix timeline sampling to cover entire video duration dynamically
  * Was limited to 40 samples (20 minutes) regardless of video length
  * Now uses ~55 samples distributed across full video duration
  * Sample interval adjusts based on video length (min 15s)
- Update prompt to emphasize reviewing ENTIRE video before selecting chapters
- Add better logging showing sample count, interval, and video duration
- Fixes issues with:
  * Chapters for unimportant small talk (e.g., weather discussion)
  * Missing chapters in second half of longer videos
  * AI only seeing beginning of video

nicolevanderhoeven committed:
Remove unused TooManyRequests import and add exception tests

- Remove unused TooManyRequests import and defensive fallback
- Add TestYouTubeAPIErrorHandling class to validate exception API contract
- Tests ensure YouTube API exceptions are importable and catchable
- Will catch breaking changes if youtube-transcript-api is upgraded

nicolevanderhoeven committed:
refactor: replace while True with explicit loop conditions and safeguards

- Add max_pages parameter (default 100) to prevent infinite loops
- Add max_retries (3) for quota exceeded errors
- Track page_count and retry_count for better control flow
- Make loop condition explicit: while page_count < max_pages
- Reset retry_count on successful requests
- Raise RuntimeError when max retries exceeded

This addresses reviewer feedback about the risks of using while True loops.

nicolevanderhoeven committed:
Refactor get_playlist_videos into smaller, more readable functions

- Extract video ID fetching logic into _fetch_playlist_video_ids
- Extract video details fetching logic into _fetch_video_details_batch
- Simplify main get_playlist_videos function to coordinate the two helpers
- Each function now has a single, clear responsibility
- Improve code readability and maintainability

nicolevanderhoeven committed:
Remove unnecessary YouTube API delays and fix retry bug

- Removed 200ms delay in _fetch_video_details_batch
- Removed 1-3 second delays in _fetch_playlist_video_ids and get_channel_videos
- Removed 2-4 second delay between playlist and channel fetches in main
- Fixed retry bug in _fetch_video_details_batch where continue would skip to next batch instead of retrying the failed one
- Added proper retry loop with max_retries tracking per batch
- 403 quota handlers now properly catch and retry after 60 second wait

nicolevanderhoeven commented on video-transcripts/transcripts.py:
```diff
@@ -1,12 +1,22 @@
+try:
+    from youtube_transcript_api._errors import TooManyRequests
+except ImportError:
+    # Fallback if TooManyRequests doesn't exist in this version
```
So actually, there wasn't a test that would have caught an error due to `TooManyRequests` not being a valid class for the API version. But in investigating this, I realised I'm not actually using `TooManyRequests` anywhere, so I've removed the import entirely.

I also added a `TestYouTubeAPIErrorHandling` test to handle errors due to the API version being upgraded.

Fixed in commit ffbda60.

nicolevanderhoeven commented on video-transcripts/transcripts.py:
Sure! Split them up into three functions (getting video IDs, fetching video details, and then another to coordinate them).

Added in c382b4d.

nicolevanderhoeven committed:
docs: clarify rate limiting delay between transcript fetches

- Add explanatory comment in transcripts.py explaining that the 2-5s delay
  prevents YouTube rate limiting (429 errors)
- Update README to accurately reflect actual delays used (2-5s, not 10-30s)
- Clarify distinction between proactive delays and reactive retry waits

nicolevanderhoeven commented on video-transcripts/transcripts.py:
```diff
@@ -115,15 +181,24 @@ def fetch_transcripts(args, videos):
+        if i < len(videos) - 1:  # Don't sleep after the last video
+            delay = random.uniform(2, 5)  # Random delay between 2-5 seconds
+            print(f"Waiting {delay:.1f} seconds before next video...")
+            time.sleep(delay)
```
Yep, this is a proactive delay to avoid hitting the YouTube rate limits. It might not be such a big deal if people are just running the script to fetch a few new videos, but I ran into rate limits a lot when regenerating the transcripts for 44 videos. I recommend keeping the delay even if it does slow down generation.

nicolevanderhoeven committed:
Simplify verbose error print statements

- Condense multiline IP blocking error messages to 2 lines
- Shorten rate limiting error messages
- Simplify XML parse error messages
- All changes reference README for detailed troubleshooting

nicolevanderhoeven committed:
Add explicit output format instructions to chapters prompt

- Instruct AI to output only the chapter list without preamble text
- Prevent conversational phrases like 'Here are the chapters' or 'Sure, I'll help'
- Addresses reviewer feedback to handle this in the prompt rather than post-processing

nicolevanderhoeven committed:
refactor: break insert_timestamps_in_transcript into smaller functions

- Extract _build_time_to_text_mapping() for building time-to-text mapping
- Extract _get_window_texts() for getting text from time windows
- Extract _extract_key_words() for keyword extraction with filtering
- Extract _calculate_line_score() for scoring line matches
- Extract _find_best_insertion_line() for finding best insertion points
- Extract _build_transcript_with_chapters() for final transcript assembly
- Main function now acts as orchestrator (36 lines, down from 113)
- Fix syntax error on line 318 (invalid character in return statement)

Benefits: improved testability, readability, and maintainability

nicolevanderhoeven commented on video-transcripts/transcripts.py:
```diff
@@ -201,6 +406,508 @@ def openai_cleanup(transcript, video_id):
+            chapters.append((seconds, timestamp_str, title))
+    return chapters
+
+def insert_timestamps_in_transcript(cleaned_transcript, chapters, raw_transcript):
```
Ugh, yeah, you're right. Fixed in 17eed88.

nicolevanderhoeven committed:
Regenerated all transcripts

danielgblanco commented:
@nicolevanderhoeven before we continue reviewing this PR, I would like to understand the core problem to solve. From experimentation it looks like the transcript cleanup process via ChatGPT can potentially solve this problem with a modification to the system prompt.

Do you think the code contributed can benefit us in the long term, as opposed to solely relying on LLMs to handle the transcript cleanup?

danielgblanco commented:
Thank you so much for all the context @nicolevanderhoeven. We (@open-telemetry/sig-end-user-maintainers) have discussed this. Our main concern was not about adding chapters, and we think that's very valuable and a great idea. However, I must personally say I had glossed over the fact that this can also help add descriptions back to YouTube so it needs to be in a specific format. So, this is then double useful!

Our concern is about having a substantial amount of code added here (larger than the current script) for the purpose of adding chapters to summaries that have already been generated via ChatGPT. The main reason we're hesitant is because we need to ensure the code that's in this repo solves issues that cannot be solved in other off-the-shelf ways.

We'd be much in favour of extending the system prompt we give ChatGPT in the cleanup, when calling `client.chat.completions.create`, maybe turn down the `temperature`. At the moment it's set to 0.7 which may be a bit too high to achieve a more deterministic result? We also wonder if switching model to `gpt-5-mini` or `gpt-5` (although this will probably reach limits on free accounts) can help.

It's clear that the non-determinism of LLMs can be an issue, but I'd argue that it's an issue in both the summaries and the chapter identification. If we can make it more deterministic by tuning parameters I think we can benefit from both summaries and chapter identification.

We don't foresee us having to re-do all transcripts every time we change this code in the future, so we also think there's an element of reviewing new video transcripts (and chapters) as they're generated in new PRs.
